src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcTransport.cs

```diff
@@ -0,0 +1,204 @@
+        /// </returns>
+        public abstract Task WaitForConnectionAsync(CancellationToken token);
+    }
+
+    internal class ServerIpcEndpoint : IpcEndpoint
+    {
+        private readonly Guid _runtimeId;
+        private readonly ReversedDiagnosticsServer _server;
+
+        public ServerIpcEndpoint(ReversedDiagnosticsServer server, Guid runtimeId)
+        {
+            _runtimeId = runtimeId;
+            _server = server;
+        }
+
+        /// <remarks>
+        /// This will block until the diagnostic stream is provided. This block can happen if
+        /// the stream is acquired previously and the runtime instance has not yet reconnected
+        /// to the reversed diagnostics server.
+        /// </remarks>
+        public override Stream Connect(TimeSpan timeout)
+        {
+            return _server.Connect(_runtimeId, timeout);
+        }
+
+        public override async Task WaitForConnectionAsync(CancellationToken token)
+        {
+            await _server.WaitForConnectionAsync(_runtimeId, token).ConfigureAwait(false);
+        }
+
+        public override bool Equals(object obj)
+        {
+            return Equals(obj as ServerIpcEndpoint);
+        }
+
+        public bool Equals(ServerIpcEndpoint other)
+        {
+            return other != null && other._runtimeId == _runtimeId && other._server == _server;
+        }
+
+        public override int GetHashCode()
+        {
+            return _runtimeId.GetHashCode() ^ _server.GetHashCode();
+        }
+    }
+
+    internal class PidIpcEndpoint : IpcEndpoint
+    {
+        public static string IpcRootPath { get; } = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? @"\\.\pipe\" : Path.GetTempPath();
+        public static string DiagnosticsPortPattern { get; } = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? @"^dotnet-diagnostic-(\d+)$" : @"^dotnet-diagnostic-(\d+)-(\d+)-socket$";
+
+        private int _pid;
+
+        /// <summary>
+        /// Creates a reference to a .NET process's IPC Transport
+        /// using the default rules for a given pid
+        /// </summary>
+        /// <param name="pid">The pid of the target process</param>
+        /// <returns>A reference to the IPC Transport</returns>
+        public PidIpcEndpoint(int pid)
+        {
+            _pid = pid;
+        }
+
+        public override Stream Connect(TimeSpan timeout)
+        {
+            string address = GetDefaultAddress();
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                var namedPipe = new NamedPipeClientStream(
+                    ".", address, PipeDirection.InOut, PipeOptions.None, TokenImpersonationLevel.Impersonation);
+                namedPipe.Connect((int)timeout.TotalMilliseconds);
+                return namedPipe;
+            }
+            else
+            {
+                var socket = new UnixDomainSocket();
+                // BUG? this path makes no use of the timeout
```

Comment:
Can't use ConnectAsync because that is a Task-returning method and the PidIpcEndpoint.Connect is non-async. However, I can probably refactor `void Connect(string path)` to take a timeout and then use Socket.BeginConnect and Socket.EndConnect internally to the method and do a WaitOne on the IAsyncResult returned from Socket.BeingConnect.

Reply:
Fixed